main/src/library/Foldable.flix

```diff
@@ -221,9 +221,41 @@ pub lawless class Foldable[t : Type -> Type] {
 
 namespace Foldable {
 
+    use Applicative.{*>};
+
     ///
     /// Returns the result of applying `combine` to all the elements in `t`, using `empty` as the initial value.
     ///
     pub def fold[t: Type -> Type, a: Type](t: t[a]): a with Foldable[t], Monoid[a] =
         Foldable.foldLeft((acc, x) -> Monoid.combine(acc, x), Monoid.empty(), t)
+
+    ///
+    /// A monadic version of `foldLeft`.
+    ///
+    /// Applies the monadic `f` to a start value `s` and all elements in `t` going
+    /// from left to right.
+    ///
+    pub def foldLeftM(f: (b, a) -> m[b] & ef, s: b, t: t[a]): m[b] & ef with Foldable[t], Monad[m] =
+        let f1 = (x, k) -> z -> Monad.flatMap(k, f(z, x));
+        s |> Foldable.foldRight(f1, x1 -> Applicative.point(x1) as & ef, t)
```

Comment:
1. What's intuition for the fact that its a foldLeft but implemented using foldRight? (is it because of CPS?)
2. Is the cast to `ef` necessary? Does it not compile without? Is it because it is in CPS?
3. Are the order of effects (like actual side-effects e.g. "print") observed "left-to-right" in the structure of t? (if t was a list say).
4. Is the reason for CPS to avoid blowing the stack?

Reply:
1. I think is is because of the CPS that `foldLeftM` is a right-fold and `foldRightM` is a left-fold. I wonder if there are implementations that build the continuation in the "natural" order? I didn't investigate when I saw the translated the Haskell code appeared stack-safe and had the side-effects in the right order.
2. Yes the cast is necessary.
3. The order of effects should be correct - the tests use a Writer-like Logger monad to show the order of printing.
4. CPS is to avoid blowing the stack - I have a test that's worked on a 2,000,000 element list.
5. Maybe, although it's easy to forget to add the non-overloaded versions Nel, Chain, etc.

Reply:
For 3. I mean a non-monadic effect, i.e. if you do `foldRightM(x -> {print(x); ....})` I would expect to see the x's in the order of the t (e.g. the list order).

Reply:
OK. This is ready, but could I ask you to add a comment to each def that explains the use of CPS and that despite it being foldLeft it is actually a foldRight and vise versa? :) (Oh and that the effects occur in the expected order). Thanks!